programmazione-concorrente.md

```diff
@@ -1 +1,437 @@
+di primitive per semplificare la programmazione concorrente.
+
+ ## Classe java.lang.Thread
+
+Per creare un thread, è sufficiente estendere la classe `java.lang.Thread`, fare l'override del metodo `run()`,
+istanziare la classe e chiamare il metodo `start()` (**NB: start, non run**):
+```java
+class MyThread extends Thread {
+    @Override
+    public void run() {
+        // Fai cose
+    }
+}
+
+...
+
+new MyThread().start(); // Fa partire il thread e ritorna immediatamente
+```
+
+Per motivi di riutilizzo del codice, viene solitamente preferito utilizzare un metodo alternativo per la creazione di
+un Thread, implementando invece l'interfaccia `java.lang.Runnable`:
+```java
+class MyRunnable implements Runnable {
+    @Override
+    public void run() {
+        // Fai cose
+    }
+}
+
+...
+
+new MyThread(new MyRunnable()).start(); // Fa partire il thread e ritorna immediatamente
+```
+
+
+Notiamo come, in entrambi i casi, il thread sia libero di accedere a qualsiasi variabile globale o passata nel costruttore,
+causando potenzialmente errori di interferenza se più thread accedono alle stesse variabili contemporaneamente.
+
+## Intrinsic lock
+
+Java, di default, associa a ogni oggetto un proprio 'lock', ovvero un oggetto che permette di implementare mutua esclusione tra
+i vari thread su sezioni di codice di interesse. Questo avviene mediante l'uso della keyword `synchronized`, specificando poi
+l'oggetto di cui si vuole utilizzare il lock:
+```java
+class Account {
+    private float balance;
+
+    public Account(float balanceIniz) { balance = balanceIniz; }
+    
+    public void deposit(float money) { 
+        synchronized(this) { 
+            balance += money; 
+        }
+    }
+
+    public void withdraw(float money) {
+        synchronized(this) {
+            balance -= money; 
+        }
+    }
+}
+```
+
+L'intrinsic lock:
+- mette in mutua esclusione sezioni di codice, permettendo a un solo thread alla volta
+  l'esecuzione, sospendendo gli altri thread chiamanti fino al completamento.
+- é un lock rientrante: se il thread, quando raggiunge una sezione critica, è già in
+  possesso del lock, continuerà la sua esecuzione senza venire sospeso.
+- quando viene acquisito/rilasciato, forza la sincronizzazione dellla cache locale del thread con 
+  la memoria centrale, garantendo che tutte le variabili lette siano aggiornate.
+
+Esiste anche una forma più breve di scrivere il codice precedente, mettendo la keyword
+direttamente sul metodo:
+```java
+class Account {
+    private float balance;
+
+    public Account(float balanceIniz) { balance = balanceIniz; }
+    
+    public synchronized void deposit(float money) { 
+        balance += money; 
+    }
+
+    public synchronized void withdraw(float money) {
+        balance -= money; 
+    }
+}
+```
+In generale, l'uso di questa forma è sconsigliato (nonostante le slide la introducano prima, mannaggia), perché:
```

Comment:
Secondo me potresti rimuovere quest'ultima parte (da riga 105 a riga 109) (nonostante condivida il disprezzo per le slide) e sostituirla con una parte dove elabori di più sul monitor pattern in cui locki un oggetto privato (magari con qualche esempio), dato che, da quel poco che ho visto, nei TDE capita di dover capire quando va bene usare l'intrinsic lock su this, quando è meglio lockare un oggetto privato e come le due cose interagiscono.
Se preferisci puoi comunque lasciare la parte commentata in modo che le fonti rimangano per chi si spulcia il markdown

Reply:
Cambiata lasciando solo la parte che effettivamente serve all'esame. In realtà non è che il discorso cambi molto, la short-hand form si può usare quando l'altra includerebbe l'intero metodo, basta fare attenzione a non buttare tutto un treno dentro alla sincronizzazione